sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelationCommand.scala

```diff
@@ -162,33 +193,36 @@ case class InsertIntoHadoopFsRelationCommand(
                 retainData = true /* already deleted */).run(sparkSession)
             }
           }
+          // The `customPartitionLocations` is getting from writing paths. Hence, it need to move
+          // files from `qualifiedOutputPath` to custom location.
+          if (updatedPartitions.nonEmpty && dynamicPartitionOverwrite) {
```

Comment:
IIUC, the goal of this pr is lazily listing partition by name using the returned modfied partition specs with dynamic partition write. Does the change related to this pr?

Reply:
Yes, the change related to this pr, because the pre-generated path has no metadata information, so all files write to the default partition path, if there is some custom partition path, we need to move the data to the custom partition inside. Maybe `customPartitionLocations.nonEmpty` is more accurate here.

Reply:
> so all files write to the default partition path, if there is some custom partition path, we need to move the data to the custom partition inside

I do not quite follow it. Before this pr, it should already be done inside commit protocol. Does this pr changes the commit protocol ? What's the relation between this method and `HadoopMapReduceCommitProtocol.commitJob` ?

Reply:
It's doesn't changes the commit protocol. Before this pr, `customPartitionLocations` is getting by pull all partitions from metastore. This pr set `customPartitionLocations`  to empty(see  https://github.com/apache/spark/pull/41628/files#diff-15b529afe19e971b138fc604909bcab2e42484babdcea937f41d18cb22d9401dR129), because it's unpredictable by lazily getting partitions from the written files when using dynamic partition write. Hence, it's need to overwrite into the `customPartitionLocations` when written files match these custom partitions.

Reply:
I see, this pr changes the output spec. But it's risky, there is a lot of history about that... After this change, there are many things chagned:
- rename partition location twice, once is at commit protocol and the other is in this method
- so the dynamic partition write staging dir changed. This pr has two staging dir since there is one more rename operation